### PR TITLE
gc: clean expired trash files, delfiles and delslices

### PR DESCRIPTION
**Pull request by Hexilee**
Enhance the gc subcommand:
- check and clean expired trash files
- check and clean the `delfiles`
- check and clean expired `delslices`

### CODE REVIEW

File: pkg/meta/base.go [hunk 1]
```diff
@@ -1363,9 +1365,8 @@ func (m *baseMeta) cleanupTrash() {
 	}
 }
 
-func (m *baseMeta) doCleanupTrash(force bool) {
+func (m *baseMeta) CleanupTrashBefore(ctx Context, edge time.Time, increProgress func()) {
```
Comment (SandyXSD) [inline pkg/meta/base.go:1368]: just export `CleanupTrash` with a limit.

File: pkg/meta/redis.go [hunk 2]
```diff
@@ -2809,53 +2809,105 @@ func (m *redisMeta) ListSlices(ctx Context, slices map[Ino][]Slice, delete bool,
...
+			}
 		}
-		return nil
+		return false, nil
 	}))
 }
 
-func (m *redisMeta) scanDeletedSlices(ctx context.Context, visitor func(s Slice) error) error {
-	if visitor == nil {
+func (m *redisMeta) scanDeletedSlices(ctx Context, scan deletedSliceScan) error {
```
Comment (SandyXSD) [inline pkg/meta/redis.go:2823]: why not use `doCleanupDelayedSlices`

**Comment by SandyXSD** [comment]
Also we'd better change the command description and printed information to reflect new changes.

File: pkg/meta/interface.go [hunk 3]
```diff
@@ -268,9 +267,11 @@ type Meta interface {
 	// ListSessions returns all client sessions.
 	ListSessions() ([]*Session, error)
 	// Statistic scan metadata by visitors.
-	Statistic(ctx context.Context, slicesDelayedScan func(Slice) error, fileDelayedScan func(ino Ino, size uint64) error) error
+	Statistic(Context, deletedSliceScan, deletedFileScan) error
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/meta/interface.go:270]: Reasonable

File: pkg/meta/redis.go (hunk 2 above)
Comment (Hexilee) [reply pkg/meta/redis.go:2823]: `doCleanupDelayedSlices` is slow and we need the progress. I would like to reuse `doCleanupDelayedSlices` and change its signature to show progress.

File: pkg/meta/interface.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/meta/interface.go:270]: Maybe we can just rename this method to `ScanDeletedObject`?

[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/meta/interface.go]: Renamed.

File: pkg/meta/redis.go (hunk 2 above)
Comment (Hexilee) [reply pkg/meta/redis.go:2823]: However, as we cannot avoid scanning all delayed slices, I think it's better to scan only once.

File: pkg/meta/base.go (hunk 1 above)
Comment (Hexilee) [reply pkg/meta/base.go:1368]: Now the `doCleanupTrash` is based on `CleanupTrashBefore`.

Comment (Hexilee) [reply pkg/meta/base.go:1368]:
```go
func (m *baseMeta) doCleanupTrash(force bool) {
	edge := time.Now().Add(-time.Duration(24*m.fmt.TrashDays+1) * time.Hour)
	if force {
		edge = time.Now()
	}
	m.CleanupTrashBefore(Background, edge, nil)
}
```

**Comment by Hexilee** [comment]
> Also we'd better change the command description and printed information to reflect new changes.

changed.

File: cmd/gc.go [hunk 4]
```diff
@@ -110,12 +110,69 @@ func gc(ctx *cli.Context) error {
...
+			delayedFileSpin.IncrInt64(int64(size))
+			if delete {
+				cleanedFileSpin.IncrInt64(int64(size))
+				return true, nil
+			}
+			return false, nil
+		},
+	)
+	if err != nil {
+		logger.Fatalf("statistic: %s", err)
```
Comment (SandyXSD) [inline cmd/gc.go:168]: Name changed.

File: pkg/meta/redis.go [hunk 5]
```diff
@@ -2809,53 +2809,105 @@ func (m *redisMeta) ListSlices(ctx Context, slices map[Ino][]Slice, delete bool,
...
+			}
+			return nil
+		})
+		close(delKeys)
+	}()
+
+	var ss []Slice
+	var rs []*redis.IntCmd
+	for key := range delKeys {
+		ss = ss[:0]
```
Comment (SandyXSD) [inline pkg/meta/redis.go:2844]: this should be within the task func because task may run several times (transaction retry)

File: cmd/gc.go (hunk 4 above)
Comment (Hexilee) [reply cmd/gc.go]: Fixed

File: pkg/meta/redis.go (hunk 5 above)
Comment (Hexilee) [reply pkg/meta/redis.go]: Fixed

**Review by SandyXSD — APPROVED** [review]

**Pull request merged**